server/src/main/java/org/elasticsearch/index/engine/ReadOnlyEngine.java

```diff
@@ -98,7 +99,24 @@ public ReadOnlyEngine(EngineConfig config, SeqNoStats seqNoStats, TranslogStats
                 indexWriterLock = obtainLock ? directory.obtainLock(IndexWriter.WRITE_LOCK_NAME) : null;
                 this.lastCommittedSegmentInfos = Lucene.readSegmentInfos(directory);
                 this.translogStats = translogStats == null ? new TranslogStats(0, 0, 0, 0, 0) : translogStats;
-                this.seqNoStats = seqNoStats == null ? buildSeqNoStats(lastCommittedSegmentInfos) : seqNoStats;
+                if (seqNoStats == null) {
+                    seqNoStats = buildSeqNoStats(lastCommittedSegmentInfos);
+                    // the global checkpoint is not always known and up to date when the engine is created,
+                    // so we only check the max seq no / global checkpoint coherency when the global checkpoint
+                    // is different from the unassigned sequence number value. In addition to that we only execute
+                    // the check if the index the engine belongs to has been created after the refactoring of the
+                    // Close Index API and its TransportVerifyShardBeforeCloseAction that guarantee that all
+                    // operations have been flushed to Lucene.
+                    final long globalCheckpoint = engineConfig.getGlobalCheckpointSupplier().getAsLong();
+                    if (globalCheckpoint != SequenceNumbers.UNASSIGNED_SEQ_NO
+                        && engineConfig.getIndexSettings().getIndexVersionCreated().onOrAfter(Version.V_6_7_0)) {
+                        if (seqNoStats.getMaxSeqNo() != globalCheckpoint) {
+                            throw new IllegalStateException("Maximum sequence number [" + seqNoStats.getMaxSeqNo()
```

Comment:
should we add an `assert false` before this line as well so that we will see this line more prominently fail on our CI? As is, it's possible that we just failover to allocating a different shard as primary, which will hide that this line tripped.

Reply:
Good idea.